## Auto Differentiation for Partial Derivatives

Determining partial derivatives by hand using rules is helpful for understanding how calculus works. In practice, however, autodiff enables us to do so more easily (especially if there are a large number of variables). For example, let's use the PyTorch automatic differentiation library to calculate the slope of $z$ with respect to both $x$ and $y$ at any given point $(x, y, z)$:

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import torch
import math

define a function $f(x, y)$ for $z = x^2 - y^2$:

In [2]:
def f(my_x, my_y):
    return my_x**2 - my_y**2

In [3]:
x = torch.tensor(0., requires_grad=True)
x

tensor(0., requires_grad=True)

In [4]:
y = torch.tensor(0., requires_grad=True)
y

tensor(0., requires_grad=True)

In [5]:
z = f(x, y) # forward pass
z

tensor(0., grad_fn=<SubBackward0>)

In [6]:
z.backward() # backward pass - autodiff

As we already know from our exercises above, the slope of the point (0, 0, 0) is zero w.r.t. both the $x$ and $y$ axes:

In [7]:
x.grad

tensor(0.)

In [8]:
y.grad

tensor(-0.)

In [9]:
# x = 3, y = 0

my_x = torch.tensor(3., requires_grad=True)
my_y = torch.tensor(0., requires_grad=True)
my_z = f(my_x, my_y) # forward pass
my_z.backward() # backward pass - autodiff
my_x.grad, my_y.grad

(tensor(6.), tensor(-0.))

In [10]:
# x = 2, y = 3

my_x = torch.tensor(2., requires_grad=True)
my_y = torch.tensor(3., requires_grad=True)
my_z = f(my_x, my_y) # forward pass
my_z.backward() # backward pass - autodiff
my_x.grad, my_y.grad

(tensor(4.), tensor(-6.))

In [11]:

# x = -2, y = -3

my_x = torch.tensor(-2., requires_grad=True)
my_y = torch.tensor(-3., requires_grad=True)
my_z = f(my_x, my_y) # forward pass
my_z.backward() # backward pass - autodiff
my_x.grad, my_y.grad

(tensor(-4.), tensor(6.))

### Partial derivatives of a cylinder's volume

The volume of a cylinder is described by $v = \pi r^2 l$ where: 

* $r$ is the radius of the cylinder
* $l$ is its length

In [12]:
def cylinder_volume(my_r, my_l):
    return math.pi * my_r**2 * my_l

In [13]:
# lets sat the radius is 3 meters
r = torch.tensor(3.).requires_grad_(True)
r

tensor(3., requires_grad=True)

In [14]:
# and the length is 5 meters
l = torch.tensor(5.).requires_grad_(True)
l       

tensor(5., requires_grad=True)

In [15]:
# the volume of the cylinder is
v = cylinder_volume(r, l)
v

tensor(141.3717, grad_fn=<MulBackward0>)

In [16]:
v.backward()
l.grad

tensor(28.2743)

As derived in the notes: $\frac{\partial v}{\partial l} = \pi r^2$

In [17]:
math.pi * 3**2

28.274333882308138

This means that with $r = 3$, a change in $l$ by one unit corresponds to a change in $v$ of 28.27 $m^3$. We can prove this to ourselves: 

In [18]:
cylinder_volume(3, 6)

169.64600329384882

In [19]:
cylinder_volume(3, 6) - cylinder_volume(3, 5)

28.274333882308127

In [20]:
cylinder_volume(3, 7) - cylinder_volume(3, 6)

28.274333882308156

For chnages in $v$ w.r.t. $r$ we have the following from the notes:

$\frac{\partial v}{\partial r}  = 2\pi rl$

In [21]:
r.grad

tensor(94.2478)

In [22]:
2 * math.pi * 3 * 5

94.24777960769379

r is included in the partial derivatives so adjusting it affects the scale of its impact on $v$. Although it's our first example in this notebook, it is typical in calculus for the derivative only to apply at an infinitesimally small $\Delta r$. The smaller the $\Delta r$, the closer to the true $\frac{\partial v}{\partial r}$


Eg: at $\Delta r = 1 \times 10^{-6}$

In [28]:
delta = 1e-6

In [29]:
(cylinder_volume(3 + delta, 5) - cylinder_volume(3, 5)) /delta # dividing by delta restores scale

94.24779531741478